File: data/compilers_comparison/src/examples/second_example/src/func21.c
#include "second_example.h" 
array_t* func21(array_t_param* vars, const unsigned long PATH0, int loopsFactor) {
   size_t pCounter = vars->size;
   array_t* array205;
   if (pCounter > 0) {
      array205 = vars->data[--pCounter];
      array205->refC++;
      DEBUG_COPY(array205->id);
   } else {
      array205 = (array_t*)malloc(sizeof(array_t));
      array205->size = 685;
      array205->refC = 1;
      array205->id = 205;
      array205->data = (unsigned int*)malloc(array205->size*sizeof(unsigned int));
      memset(array205->data, 0, array205->size*sizeof(unsigned int));
      DEBUG_NEW(array205->id);
   }
   array_t_param params0;
   params0.size = 1;
   params0.data = (array_t**)malloc(params0.size*sizeof(array_t*));
   params0.data[0] = array205;
   array_t* array206 = func23(&params0, loopsFactor);
   DEBUG_RETURN(array206->id);
   free(params0.data);
   if(PATH0 & 1) {
      array_t* array207;
      if (pCounter > 0) {
         array207 = vars->data[--pCounter];
         array207->refC++;
         DEBUG_COPY(array207->id);
      } else {
         array207 = (array_t*)malloc(sizeof(array_t));
         array207->size = 176;
         array207->refC = 1;
         array207->id = 207;
         array207->data = (unsigned int*)malloc(array207->size*sizeof(unsigned int));
         memset(array207->data, 0, array207->size*sizeof(unsigned int));
         DEBUG_NEW(array207->id);
      }
      unsigned int loop60 = 0;
      unsigned int loopLimit60 = (rand()%loopsFactor)/3 + 1;
      for(; loop60 < loopLimit60; loop60++) {
         for (int i = 0; i < array205->size; i++) {
            array205->data[i]--;
         }
         for (int i = 0; i < array207->size; i++) {
            if (array207->data[i] == 59) { 
               return array207;
            }
         }
      }
      array_t_param params1;
      params1.size = 3;
      params1.data = (array_t**)malloc(params1.size*sizeof(array_t*));
      params1.data[0] = array205;
      params1.data[1] = array206;
      params1.data[2] = array207;
      array_t* array208 = func36(&params1, rng(), loopsFactor);
      DEBUG_RETURN(array208->id);
      free(params1.data);
      array208->refC--;
      if(array208->refC == 0) {
         free(array208->data);
         free(array208);
         DEBUG_FREE(array208->id);
      }
      array207->refC--;
      if(array207->refC == 0) {
         free(array207->data);
         free(array207);
         DEBUG_FREE(array207->id);
      }
   }
   else {
      array_t_param params1;
      params1.size = 2;
      params1.data = (array_t**)malloc(params1.size*sizeof(array_t*));
      params1.data[0] = array205;
      params1.data[1] = array206;
      array_t* array209 = func32(&params1, rng(), loopsFactor);
      DEBUG_RETURN(array209->id);
      free(params1.data);
      array209->refC--;
      if(array209->refC == 0) {
         free(array209->data);
         free(array209);
         DEBUG_FREE(array209->id);
      }
   }
   array205->refC--;
   if(array205->refC == 0) {
      free(array205->data);
      free(array205);
      DEBUG_FREE(array205->id);
   }
   return array206;
}

File: data/multilang_comparison/program_nim/src/func21.nim
proc func21*(vars: ptr ArrayTParam, PATH0: uint64, loopsFactor: cint): ptr ArrayT =
   var pCounter = vars.size
   var array237: ptr ArrayT
   if pCounter > 0:
      dec pCounter
      array237 = vars.data[pCounter]
      inc array237.refC
      debugCopy(array237.id)
   else:
      array237 = cast[ptr ArrayT](alloc(sizeof(ArrayT)))
      array237.size = 231
      array237.refC = 1
      array237.id = 237
      array237.data = cast[ptr UncheckedArray[uint32]](alloc(array237.size * sizeof(uint32)))
      zeroMem(array237.data, array237.size * sizeof(uint32))
      debugNew(array237.id)
   var params0: ArrayTParam
   params0.size = 1
   params0.data = cast[ptr UncheckedArray[ptr ArrayT]](alloc(params0.size * sizeof(ptr ArrayT)))
   params0.data[0] = array237
   var array238: ptr ArrayT = func23(addr params0, loopsFactor)
   debugReturn(array238.id)
   dealloc(params0.data)
   if (PATH0 and 1'u64) != 0:
      discard
      var array239: ptr ArrayT
      if pCounter > 0:
         dec pCounter
         array239 = vars.data[pCounter]
         inc array239.refC
         debugCopy(array239.id)
      else:
         array239 = cast[ptr ArrayT](alloc(sizeof(ArrayT)))
         array239.size = 404
         array239.refC = 1
         array239.id = 239
         array239.data = cast[ptr UncheckedArray[uint32]](alloc(array239.size * sizeof(uint32)))
         zeroMem(array239.data, array239.size * sizeof(uint32))
         debugNew(array239.id)
      let loopLimit69 = (50) div (3) + 1
      for loop69 in 0 ..< loopLimit69:
         discard
         for i in 0 ..< array239.size:
            dec array239.data[i]
         for i in 0 ..< array238.size:
            if array238.data[i] == 18:
               return array238
      var params1: ArrayTParam
      params1.size = 3
      params1.data = cast[ptr UncheckedArray[ptr ArrayT]](alloc(params1.size * sizeof(ptr ArrayT)))
      params1.data[0] = array237
      params1.data[1] = array238
      params1.data[2] = array239
      var array240: ptr ArrayT = func36(addr params1, getPath(), loopsFactor)
      debugReturn(array240.id)
      dealloc(params1.data)
      dec array240.refC
      if array240.refC == 0:
         dealloc(array240.data)
         dealloc(array240)
         debugFree(array240.id)
      dec array239.refC
      if array239.refC == 0:
         dealloc(array239.data)
         dealloc(array239)
         debugFree(array239.id)
   else:
      discard
      var params1: ArrayTParam
      params1.size = 2
      params1.data = cast[ptr UncheckedArray[ptr ArrayT]](alloc(params1.size * sizeof(ptr ArrayT)))
      params1.data[0] = array237
      params1.data[1] = array238
      var array241: ptr ArrayT = func32(addr params1, getPath(), loopsFactor)
      debugReturn(array241.id)
      dealloc(params1.data)
      dec array241.refC
      if array241.refC == 0:
         dealloc(array241.data)
         dealloc(array241)
         debugFree(array241.id)
   dec array237.refC
   if array237.refC == 0:
      dealloc(array237.data)
      dealloc(array237)
      debugFree(array237.id)
   return array238

